pyo3-macros-backend/src/method.rs

```diff
@@ -455,13 +455,27 @@ impl<'a> FnSpec<'a> {
         let func_name = &self.name;
 
         let rust_call = |args: Vec<TokenStream>| {
-            let call = quote! { function(#self_arg #(#args),*) };
-            let wrapped_call = if self.asyncness.is_some() {
-                quote! { _pyo3::PyResult::Ok(_pyo3::impl_::wrap::wrap_future(#call)) }
-            } else {
-                quotes::ok_wrap(call)
-            };
-            quotes::map_result_into_ptr(wrapped_call)
+            let mut call = quote! { function(#self_arg #(#args),*) };
+            if self.asyncness.is_some() {
+                let python_name = &self.python_name;
+                let qualname = match cls {
+                    Some(cls) => quote! {
+                        _pyo3::impl_::coroutine::method_coroutine_qualname::<#cls>(py, stringify!(#python_name))
+                    },
+                    None => quote! {
+                        _pyo3::impl_::coroutine::coroutine_qualname(py, py.from_borrowed_ptr_or_opt::<_pyo3::types::PyModule>(_slf), stringify!(#python_name))
+                    },
+                };
+                call = quote! {{
+                    let future = #call;
+                    _pyo3::impl_::coroutine::new_coroutine(
+                        _pyo3::types::PyString::new(py, stringify!(#python_name)).into(),
```

Comment:
It would probably be worth interning here using our `intern!` macro, and changing `new_coroutine` to take `N: IntoPy<Py<PyString>>`.